docs/src/capi/explanation/in-place-upgrades.md

```diff
@@ -0,0 +1,236 @@
+with `upgrade-to: channel=1.31/stable`. The single machine upgrader 
+(which is watching for changes on machines), notices this annotation 
+and attempts to upgrade the Kubernetes version of that machine to the 
+specified version.
+
+Because {{product}} is shipped in a snap package, performing an upgrade 
+can be as easy as doing a `snap refresh`. Upgrade methods or options 
+can be specified to upgrade to a snap channel, revision, or install a 
+new snap from a file already placed on the machine to make up for 
+air-gapped environments.
+
+When the upgrade is finished successfully, we will notice (at least) the 
+following annotations on the machine:
+
+```yaml
+annotations:
+  v1beta2.k8sd.io/in-place-upgrade-release: "channel=1.31/stable"
+  v1beta2.k8sd.io/in-place-upgrade-status: "done"
+```
+
+If the upgrade fails, the controller will mark the machine with the 
+following annotations and retry immediately:
+
+```yaml
+annotations:
+  # the `upgrade-to` causes the retry to happen
+  v1beta2.k8sd.io/in-place-upgrade-to: "channel=1.31/stable"
+  v1beta2.k8sd.io/in-place-upgrade-status: "failed"
+
+  # orchestrator will notice this annotation and knows that the 
+  # upgrade for this machine failed
+  v1beta2.k8sd.io/in-place-upgrade-last-failed-attempt-at: "Sat, 7 Nov 
+  2024 13:30:00 +0400"
+```
+
+For the detailed implementation of this controller, make sure to 
+checkout the [{{product}} CAPI repository][capi-repo].
+
+Overall, by applying and removing annotations, the single machine 
+upgrader determines the upgrade status of the machine it’s trying to 
+reconcile and takes necessary actions to successfully complete an 
+in-place upgrade. The following diagram shows the flow of the in-place 
+upgrade of a single machine:
+
+```{figure} ../../assets/in-place-upgrades/single-machine.png
+   :width: 100%
+   :alt: Single Machine In-Place Upgrade Flow
+
+   Single Machine In-Place Upgrade Flow
+```
+
+### Machine Upgrade Process
+
+{{product}} has a daemon running in the background called the `k8sd`. 
+It’s responsible for many things in the context of {{product}} but 
+one of them is to expose certain endpoints that can be used to 
+interact with the cluster. The single machine upgrader calls the 
+`/snap/refresh` endpoint on the machine that it’s trying to upgrade. 
+This endpoint call will trigger the “actual” upgrade process and in 
+the meantime, the `/snap/refresh-status` will be called periodically 
+by the single machine upgrader to see how things are going. It’s worth 
+noting that ensuring secure communication between the single machine 
+upgrader and the {{product}} daemon (k8sd) is really important and 
+something that is handled internally as well.
+
+```{figure} ../../assets/in-place-upgrades/k8sd-call.png
+   :width: 100%
+   :alt: Single Machine In-Place Upgrade Controller Calling The `k8sd` 
+   Endpoints
+
+   Single Machine In-Place Upgrade Controller Calling The `k8sd` Endpoints
+```
```

Comment:
This is more an implementation detail, does the user really need to know this?

Reply:
Maybe I'm wrong, but when I think of the audience of an explanation I'm not necessarily thinking about someone that is reading this "just to understand how to work *with* it, or how to operate this product". I think that is what a "How to" is for. I think of an "Explanation" as something that will help people understand how and why this thing is working underneath, something like a _can you tell me about...?._

Quoting from [Diataxes](https://www.diataxis.fr/explanation/):
> Explanation deepens and broadens the reader’s understanding of a subject. It brings clarity, light and context.
>
> Explanation - understanding oriented, theoretical knowledge, that serves our study
> The concept of reflection is important. Reflection occurs after something else, and depends on something else, yet at the same time brings something new - shines a new light - on the subject matter.
>
> The perspective of explanation is higher and wider than that of the other not types. It does not take the user’s eye-level view, as in a how-to guide, or a close-up view of the machinery, like reference material. Its scope in each case is a topic - “an area of knowledge”, that somehow has to be bounded in a reasonable, meaningful way.
> For the user, explanation joins things together. It’s an answer to the question: Can you tell me about …?
>
> It’s documentation that it makes sense to read while away from the product itself (one could say, explanation is the only kind of documentation that it might make sense to read in the bath).

Reply:
And I would argue that this implementation detail is very unlikely to change. I mean the exact endpoints _might_ (e.g. `/snap/refresh-status`), but I don't think the `k8sd` is going anywhere. So I believe this won't rot as well.